pkg/api/types.go

```diff
@@ -2119,34 +2119,121 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
-// Type and constants for component health validation.
+// Type and constants for component type.
+type ComponentType string
+
+// ComponentType constants define some known types of components.
+// Each type may have multiple instances, as long as their names are unique.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
+}
+
+type ComponentPhase string
+
+// These are the valid phases of a component.
+const (
+	// ComponentPending means the component has been created, but is still starting up.
+	ComponentPending ComponentPhase = "Pending"
+	// ComponentRunning means the component has started, but may or may not be healthy.
+	ComponentRunning ComponentPhase = "Running"
+	// ComponentTerminated means the component has been removed from the cluster, whether cleanly or not.
+	ComponentTerminated ComponentPhase = "Terminated"
```

Comment:
It's generally a bad idea to remove records from a db row. A "deleted" column is usually used for a number of reasons, usually for reporting, metrics, or cause. Of course, this eventually requires clean up to remove "deleted" records, usually after they've been processed or after another timeout. 

In this case Terminated expresses the expectation that the Component is not coming back.

An alternative method would be to delete the component record but keep a log of events/updates/transitions.

In this case, a Terminated Component can have the following Conditions:
- Terminated Cleanly
  - Sent a final status update before exiting
  - Reported a Reason (e.g. SIGTERM, SIGINT)
- Terminated Uncleanly
  - Sent a final status update before exiting OR marked as terminated by timeout.
  - Reported a Reason (e.g. error, timeout)